SwiftReflector/MarshalEngine.cs

```diff
@@ -944,6 +950,44 @@ CSBaseExpression MarshalProtocolListAsPointer (BaseDeclaration typeContext, CSPa
 			}
 		}
 
+		CSBaseExpression MarshalAssociatedTypePathAsPointer (BaseDeclaration typeContext, FunctionDeclaration funcDecl, CSParameter p, NamedTypeSpec pointsTo, bool isReturnValue)
+		{
+			// in swift, we have a type which is T.Assoc
+			// in C#, it's going to be an unadorned generic, ATAssoc
+			// byte *pBufferPtr = stackalloc byte[StructMarshal.Strideof (typeof(ATAssoc))];
+			// IntPtr pBufferIntPtr = new IntPtr (pBuffer);
+			// if !isReturnValue:
+			//      StructMarshal.Marshaler.ToSwift (typeof(ATAssoc), pBufferIntPtr);
+			// SomePInvoke (pBufferIntPtr);
+			// p = StructMarshal.Marshaler.ToNet<ATAssoc>(pBufferIntPtr);
+			RequiredUnsafeCode = true;
+
+			use.AddIfNotPresent (typeof (StructMarshal));
+
+			var bufferIdent = new CSIdentifier (Uniqueify (p.Name.Name + "Buffer", identifiersUsed));
+			identifiersUsed.Add (bufferIdent.Name);
+			var retvalbufferDecl = CSVariableDeclaration.VarLine (
+				CSSimpleType.ByteStar, bufferIdent,
+				CSArray1D.New (CSSimpleType.Byte, true, new CSFunctionCall ("StructMarshal.Marshaler.Strideof", false, p.CSType.Typeof ())));
+			preMarshalCode.Add (retvalbufferDecl);
+
+			var bufferPtrIdent = new CSIdentifier (Uniqueify (p.Name.Name + "BufferPtr", identifiersUsed));
+			identifiersUsed.Add (bufferPtrIdent.Name);
+			var retvalBufferPtrDecl = CSVariableDeclaration.VarLine (
+				CSSimpleType.IntPtr, bufferPtrIdent,
+				new CSFunctionCall ("IntPtr", true, bufferIdent));
+			preMarshalCode.Add (retvalBufferPtrDecl);
+
+			if (!isReturnValue) {
+				preMarshalCode.Add (CSFunctionCall.FunctionCallLine ("StructMarshal.Marshaler.ToSwift", false,
+					p.CSType.Typeof (), p.Name, bufferPtrIdent));
+			}
+			var toNetCall = new CSFunctionCall (String.Format ("StructMarshal.Marshaler.ToNet<{0}>", p.CSType.ToString ()),
+				  false, bufferPtrIdent, CSConstant.Val (true));
+			postMarshalCode.Add (new CSLine (toNetCall));
+			return bufferPtrIdent;
```

Comment:
If I am understanding this code (doubtful...), the return value is a `CSIdentifier` that stores a unique string to identify the C# function call that is created here.

Reply:
You're close.
Most every function of the form `MarshalSomeType` returns a `CSBaseExpression` which represents an expression that gets passed into swift to represent the parameter `p`. In some cases, it's easy. If the type of `p` is a scalar, I can return `p` as an identifier, since that goes into swift with no changes.
If the the of `p` is `ISwiftObject` then I pass in `p.SwiftObject` which is the handle to underlying swift object.
In the case of an associated type path (which is a generic), there is pre-marshaling code which in a typical case:
1 - stack allocates a buffer the size of the swift type the `p` represents
2 - does runtime marshaling to put the appropriate value of that buffer
3 - makes an `IntPtr` from that buffer which will get passed in to the swift function
Upon return from calling the swift function, nothing else happens.
Now because I'm super lazy, if the return value of a function won't necessarily match our register return calling conventions, I do a trick where I inject the return value as a fake C# parameter with a flag saying "this is a return value". It acts almost the same as the parameter case but with two differences:
1 - it doesn't do runtime marshaling to copy into the buffer because there is no value
2 - it adds post marshaling code to put unmarshal the return value from the buffer into something C#-ish

Reply:
@stephen-hawley Thanks for the explanation! The return value special case is particularly mind-bendy. 👍